### DISCUSSION TITLE
How can I edit the thread's metadata outside of React paradigm?

### DISCUSSION THREAD

**Discussion by xmliszt** [liveblocks/liveblocks#1620] [Help] [answered]
In my project, I am trying to integrate Liveblocks with TLDraw. I want to update the thread's metadata for example the x and y position, whenever the translation of the thread shape ends, which is provided as a function in TLDraw's shape util class. I can't use `useEditMetadata()` here because the hook needs to be inside a functional component. Is there a way I can somehow hook the update into the `onTranslateEnd` callback that exists in the TLDraw class (outside of React paradigm)?

**Comment by CTNicholas** [comment, marked as answer]
Hey @xmliszt, we don't currently offer a way to edit threads directly on the client outside of React, however it is possible to do this on the server if you call an endpoint using [`liveblocks.editThreadMetadata`](https://liveblocks.io/docs/api-reference/liveblocks-node#post-rooms-roomId-threads-threadId-metadata).

```tsx
const editedMetadata = await liveblocks.editThreadMetadata({
  roomId: "my-room-id",
  threadId: "th_d75sF3...",

  data: {
    metadata: {
      x: 100, 
      y: 200
    },
    userId: "marc@example.com",
    updatedAt: new Date(), // Optional
  },
});

// { x: 100, y: 200 }
console.log(editedMetadata);
```

**Reply by xmliszt** [reply]
Thank you @CTNicholas ! In the end, I managed to figure out a way together with the use of Liveblocks Storage! So basically:

Liveblocks Storage: manage syncing thread components' x / y / w / h properties
Liveblocks Threads: manage syncing of the actual thread contents (comments, resolve, reactions, etc.)

In this way, I can use the hooks in my custom React component, with a provider that uses `useThreads()` to get the threads and provides a convenient function `findThreadById(id)` for each custom thread component to consume to get the thread data!

But thanks for the help anyway as it inspired me to arrive at this solution!

**Reply by CTNicholas** [reply]
Ah, yes! That will work well, and be even more real-time, good idea. 

Let us know if we can help more!

**Comment by CTNicholas** [comment]
> Thank you @CTNicholas ! In the end, I managed to figure out a way together with the use of Liveblocks Storage! So basically:
> 
> Liveblocks Storage: manage syncing thread components' x / y / w / h properties Liveblocks Threads: manage syncing of the actual thread contents (comments, resolve, reactions, etc.)
> 
> In this way, I can use the hooks in my custom React component, with a provider that uses `useThreads()` to get the threads and provides a convenient function `findThreadById(id)` for each custom thread component to consume to get the thread data!
> 
> But thanks for the help anyway as it inspired me to arrive at this solution!
